tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc

```diff
@@ -30,6 +31,63 @@ namespace mlir {
 namespace disc_ral {
 namespace {
 
+// Returns true if a tensor is constant and all its elements are equal to
+// `elemVal`
+bool allElementsAreSameValue(Value tensor, int64_t elemVal) {
+  auto definingOp = tensor.getDefiningOp();
+  if (!definingOp) return false;
+  if (isa<mhlo::BroadcastOp, mhlo::BroadcastInDimOp,
+          mhlo::DynamicBroadcastInDimOp>(definingOp)) {
+    return allElementsAreSameValue(definingOp->getOperand(0), elemVal);
+  }
+  DenseElementsAttr denseAttr;
+  if (!matchPattern(tensor, m_Constant(&denseAttr))) return false;
+  if (denseAttr.getNumElements() != 1 && !denseAttr.isSplat()) return false;
+
+  Type elemTy = denseAttr.getElementType();
+  if (elemTy.isIntOrIndex()) {
+    return (*denseAttr.getValues<APInt>().begin()).getSExtValue() == elemVal;
+  } else if (elemTy.isa<FloatType>()) {
+    return (*denseAttr.getValues<APFloat>().begin()).convertToDouble() ==
+           elemVal;
+  }
+  return false;
+}
+
+// `x + 0` or `0 + x` could be simplified to `x`
+struct AddZeroTensorOp : public OpRewritePattern<mhlo::AddOp> {
```

Comment:
How about adding patterns for `x - 0` and something similar.

Reply:
It may be added in the future.  However, some similar patterns may not be safe. For example, `x * 0 = 0` is not true if `x` is `NaN`.